#Trabalho prático # 5
##Introdução à Computação Numérica

###Tópico: Solução de Problemas de Valor Inicial (PVI)

---

Nome: *Eduarda Leal*

DRE: *118199369*


---

**Instruções:**

**Preencha o notebook** com seus dados, seus códigos e **rode todos os códigos**

**Inclua  um relatório, na parte final do notebook,** com suas conclusões, respostas, observações e comentários relacionados com os tópicos e questões abordadas nas tarefas.  

**A elaboração dos códigos pode ser realizada em duplas, mas o resto do trabalho é individual.**

*Se você fez o seu trabalho em dupla inclua essa informação no relatório. Além disso, se você  também recebeu ajuda de um(a) colega/amigo(a) ou consultou/usou informações da internet ou outras fontes, coloque referências à elas no seu relatório.*


**Entrega do trabalho:**

*Data de entrega: 12/julho/2023*

---


## Tarefa #1:

 Desenvolver códigos, na forma de funções que implementem:

a) O Método de Euler.

b) O Método  de Euler aperfeiçoado (ou Método de Heun).

c) O Método de Runge-Kutta clássico de quarta ordem.



In [ ]:
# Letra A

# Implementação do método de Euler
def euler(f, x0, y0, h, n):
    x = x0
    y = y0
    result = [(x, y)]

    for _ in range(n):
        # Calcula o próximo valor de y usando a fórmula do método de Euler
        y = y + h * f(x, y)
        # Atualiza o valor de x
        x = x + h
        # Adiciona o par (x, y) ao resultado
        result.append((x, y))

    return result


# Letra B

# Implementação do método de Euler aperfeiçoado
def euler_aperfeicoado(f, x0, y0, h, n):
    x = x0
    y = y0
    result = [(x, y)]

    for _ in range(n):
        # Calcula as inclinações k1 e k2 usando a fórmula do método de Euler aperfeiçoado
        k1 = h * f(x, y)
        k2 = h * f(x + h, y + k1)
        # Calcula o próximo valor de y usando a fórmula do método de Euler aperfeiçoado
        y = y + 0.5 * (k1 + k2)
        # Atualiza o valor de x
        x = x + h
        # Adiciona o par (x, y) ao resultado
        result.append((x, y))

    return result


# Letra C

# Implementação do método de Runge-Kutta de quarta ordem
def runge_kutta(f, x0, y0, h, n):
    x = x0
    y = y0
    result = [(x, y)]

    for _ in range(n):
        # Calcula as inclinações k1, k2, k3 e k4 usando a fórmula do método de Runge-Kutta
        k1 = h * f(x, y)
        k2 = h * f(x + (1/2) * h, y + (1/2) * k1)
        k3 = h * f(x + (1/2) * h, y + (1/2) * k2)
        k4 = h * f(x + h, y + k3)

        # Calcula o próximo valor de y usando a fórmula do método de Runge-Kutta
        y = y + (k1 + 2*k2 + 2*k3 + k4) / 6
        # Atualiza o valor de x
        x += h
        # Adiciona o par (x, y) ao resultado
        result.append((x, y))

    return result



## Tarefa #2: Paraquedista em queda livre

Utilize os seus códigos (da tarefa anterior, itens a), b) e c) ) para responder a seguinte questão.   
**Qual será a velocidade vertical atingida  por um paraquedista em queda livre (a velocidade inicial é nula), $25\,s$ após o salto?**


a) Considere que a taxa de variação instantânea da velocidade do paraquedista é dada pela EDO
$$\frac{dv}{dt}=g-\frac{c}{m} v, $$
onde $g=9.81\,m/s^2$ é a aceleração da gravidade, $m=75\,kg$ é a massa do paraquedista e $c=12.5\,kg/s$ é o coeficiente de arrasto correspondente ao paraquedas.

Para cada método implementado determine um número de passos tal que o erro na aproximação da velocidade final $v(25)$ seja inferior a $10^{-2}$, se sabemos que o valor exato da velocidade  pode ser calculado pela fórmula
$$v(t) = \frac{gm}{c}\left( 1-e^{-(c/m)t}\right).$$

Dica: Isto poderá ser feito por tentativa e erro.

b) Considere uma outra lei para a taxa de variação instantânea da velocidade do paraquedista dada pela EDO
$$\frac{dv}{dt}=g-\frac{C(v)}{m} v, $$
onde a função empiríca $C(v)$ que caracteriza a resistência do paraquedas é dada por
$$C(v) = c\left[1+\alpha \left(\frac{v}{v_{max}}\right)^\beta\right],$$
em que $c=12.5\,kg/s$, $\alpha=0.2$, $\beta =1.1$ e $v_{max} = 46\,m/s$.  (Observe que no caso $\alpha=0$ este modelo coincide com o apresentado no item a).)

Obtenha a velocidade final usando cada um dos métodos implementados considerando o mesmo número de passos definido no item a)).

c) Compare as velocidades obtidas nos itens a) e b).


In [ ]:
#Letra A

import math

g = 9.81  # Aceleração da gravidade em m/s^2
m = 75  # Massa do paraquedista em kg
c = 12.5  # Coeficiente de arrasto em kg/s

def f(t, v):
    return g - (c/m) * v

#Função para encontrar número de passos necessários para que erro seja menor que 10^-2.
def qnt_passos(metodo):
    n = 1
    while True:
        resultado = metodo(f, 0, 0, 25/n, n)
        aprox_velocidade = resultado[-1][1]
        velocidade_exata = (g * m / c) * (1 - math.exp(-(c / m) * 25))
        erro = abs(aprox_velocidade - velocidade_exata)
        if erro < 10**(-2):
            return n, erro, resultado
        n += 1

n_euler, erro_euler, euler_resultado = qnt_passos(euler)
n_euler_aprimorado, erro_euler_aprimorado, euler_aprimorado_resultado = qnt_passos(euler_aperfeicoado)
n_runge_kutta, erro_runge_kutta, runge_kutta_resultado = qnt_passos(runge_kutta)

velocidade_euler_A = euler_resultado[-1][1]
velocidade_euler_aprimorado_A = euler_aprimorado_resultado[-1][1]
velocidade_runge_kutta_A = runge_kutta_resultado[-1][1]

print(f"Método de Euler\nNúmero de passos: {n_euler}\nVelocidade aproximada: {velocidade_euler_A:.4f} m/s Erro de aproximação: {erro_euler:.4f}\n")
print(f"Método de Euler aperfeiçoado\nNúmero de passos: {n_euler_aprimorado}\nVelocidade aproximada: {velocidade_euler_aprimorado_A:.4f} m/s Erro de aproximação: {erro_euler_aprimorado:.4f}\n")
print(f"Método de Runge-Kutta\nNúmero de passos: {n_runge_kutta}\nVelocidade aproximada: {velocidade_runge_kutta_A:.4f} m/s Erro de aproximação: {erro_runge_kutta:.4f}")



Método de Euler
Número de passos: 791
Velocidade aproximada: 57.9574 m/s Erro de aproximação: 0.0100

Método de Euler aperfeiçoado
Número de passos: 35
Velocidade aproximada: 57.9376 m/s Erro de aproximação: 0.0099

Método de Runge-Kutta
Número de passos: 7
Velocidade aproximada: 57.9409 m/s Erro de aproximação: 0.0066


In [ ]:
#Letra B

c = 12.5  # Coeficiente de arrasto em kg/s
alfa = 0.2
beta = 1.1
v_max = 46  # m/s

def C(v):
    return c * (1 + alfa * (v/v_max)**beta)

def f(t, v):
    return g - (C(v)/m) * v

euler_resultado = euler(f, 0, 0, 25/n_euler, n_euler)
euler_aprimorado_resultado = euler_aperfeicoado(f, 0, 0, 25/n_euler_aprimorado, n_euler_aprimorado)
runge_kutta_resultado = runge_kutta(f, 0, 0, 25/n_runge_kutta, n_runge_kutta)

velocidade_euler_B = euler_resultado[-1][1]
velocidade_euler_aprimorado_B = euler_aprimorado_resultado[-1][1]
velocidade_runge_kutta_B = runge_kutta_resultado[-1][1]

print(f"Velocidade final pelo Método de Euler:{velocidade_euler_B:.4f} m/s")
print(f"Velocidade final pelo Método de Euler aperfeiçoado: {velocidade_euler_aprimorado_B:.4f} m/s")
print(f"Velocidade final pelo Método de Runge-Kutta:{velocidade_runge_kutta_B:.4f} m/s")

Velocidade final pelo Método de Euler:48.4166 m/s
Velocidade final pelo Método de Euler aperfeiçoado: 48.4091 m/s
Velocidade final pelo Método de Runge-Kutta:48.4066 m/s


In [ ]:
#Letra C

print(f"Diferença de velocidade no Método de Euler: {velocidade_euler_A - velocidade_euler_B:.4f}")
print(f"Diferença de velocidade no Método de Euler aprimorado: {velocidade_euler_aprimorado_A - velocidade_euler_aprimorado_B:.4f}")
print(f"Diferença de velocidade no Método de Runge-Kutta: {velocidade_runge_kutta_A - velocidade_runge_kutta_B:.4f}")



Diferença de velocidade no Método de Euler: 9.5409
Diferença de velocidade no Método de Euler aprimorado: 9.5285
Diferença de velocidade no Método de Runge-Kutta: 9.5343


## Tarefa #3: Propagação de uma doença infecciosa

Para modelar a dinâmica de uma doença infecciosa em que os indivíduos se recuperam sem imunidade, é usado o seguinte sistema de duas EDOs conhecido como modelo SIS:

$$
\left\{
\begin{aligned}
\frac{d S}{dt} &= -\beta S I + \gamma I\\
\frac{d I}{dt} &= \beta S I - \gamma I
\end{aligned}
\right.
$$

onde $S$ e $I$ representam o número de indivíduos suscetíveis e infectados, respectivamente, $t$ o tempo,  $\beta$ a velocidade de contacto ou infecção da doença e $\gamma$ a taxa de recuperação média.

Considere uma cidade com uma população de $10000$ indivíduos todos suscetíveis. Se um único indivíduo infectado entra na cidade to tempo $t = 0$, qual será a quantidade de indivíduos suscetíveis e infectados após $3$ semanas? Use os parâmetros $\beta = 0.0003/(\mathit{indivíduo\cdot dia})$ e $\gamma = 0.15/\mathit{dia}$.

Utilize o seu código do método de RK de $4^a$ ordem (tarefa #1, item c)) para responder a essa questão. Procure uma aproximação com erro estimado inferior a $0.05$. Explique.

In [ ]:
import numpy as np
import matplotlib.pyplot as plt

def SIS_model(x, y):
    S, I = y
    beta = 0.0003  # taxa de contato ou infecção da doença
    gamma = 0.15  # taxa de recuperação média

    dSdt = -beta * S * I + gamma * I
    dIdt = beta * S * I - gamma * I

    return np.array([dSdt, dIdt])

# Condições iniciais
S0 = 10000
I0 = 1
y0 = np.array([S0, I0])

# Parâmetros de tempo
t0 = 0  # tempo inicial
tf = 21  # tempo final em dias
dt = 1  # passo de tempo em dias
n = int((tf - t0) / dt)  # número de iterações

# Integração numérica usando o método de Runge-Kutta
results = runge_kutta(SIS_model, t0, y0, dt, n)

# Extração dos resultados
t_list = [result[0] for result in results]
S_list = [result[1][0] for result in results]
I_list = [result[1][1] for result in results]

# Resultados após 3 semanas
final_S = S_list[-1]
final_I = I_list[-1]

print("Quantidade de indivíduos suscetíveis após 3 semanas:", int(final_S))
print("Quantidade de indivíduos infectados após 3 semanas:", int(final_I))

Quantidade de indivíduos suscetíveis após 3 semanas: 2061
Quantidade de indivíduos infectados após 3 semanas: 7939


# Relatório

## Tarefa 1

Os códigos da implementação dos métodos se encontram na primeira seção (Tarefa 1)

## Tarefa 2

**LETRA A**

Analisando os resultados obtidos abaixo, podemos perceber que o método Runge-Kutta é mais eficiente pois obteve menor erro na aproximação e com menor quantidade de passos.

Resultados obtidos:

**Método de Euler**

Número de passos: 791

Velocidade aproximada: 57.9574 m/s

Erro de aproximação: 0.0100

**Método de Euler aperfeiçoado**

Número de passos: 35

Velocidade aproximada: 57.9376 m/s

Erro de aproximação: 0.0099

**Método de Runge-Kutta**

Número de passos: 7

Velocidade aproximada: 57.9409 m/s

Erro de aproximação: 0.0066

**LETRA B**

Velocidade adicionando a resistência do paraquedas:

Velocidade final pelo Método de Euler: 48.4166 m/s

Velocidade final pelo Método de Euler aperfeiçoado: 48.4091 m/s

Velocidade final pelo Método de Runge-Kutta: 48.4066 m/s

**LETRA C**

Calculando a diferença da velocidade encontrada na letra a e b:

Diferença de velocidade no Método de Euler: 9.5409

Diferença de velocidade no Método de Euler aprimorado: 9.5285

Diferença de velocidade no Método de Runge-Kutta: 9.5343

## Tarefa 3

Os resultados encontrados após 3 semanas de simulação mostraram que a dinâmica da doença na população mudou drasticamente. Inicialmente, com todos os indivíduos suscetíveis e apenas um indivíduo infectado, observamos uma rápida disseminação da doença com um aumento significativo no número de indivíduos infectados. Isso ocorre porque a taxa de contato é maior que a taxa de recuperação.

Após 3 semanas, a quantidade de indivíduos suscetíveis será de 2061 e de indivíduos infectados será de 7939.


